internal/topology/variables/cluster_variable_validation.go

```diff
@@ -0,0 +1,124 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package variables
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"k8s.io/apiextensions-apiserver/pkg/apis/apiextensions"
+	"k8s.io/apiextensions-apiserver/pkg/apiserver/validation"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+)
+
+// ValidateClusterVariables validates clusterVariable via the schemas in the corresponding clusterClassVariable.
+func ValidateClusterVariables(clusterVariables []clusterv1.ClusterVariable, clusterClassVariables []clusterv1.ClusterClassVariable, fldPath *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+
+	// Build maps for easier and faster access.
+	clusterVariablesMap := map[string]*clusterv1.ClusterVariable{}
+	for i := range clusterVariables {
+		clusterVariablesMap[clusterVariables[i].Name] = &clusterVariables[i]
+	}
+	clusterClassVariablesMap := map[string]*clusterv1.ClusterClassVariable{}
+	for i := range clusterClassVariables {
+		clusterClassVariablesMap[clusterClassVariables[i].Name] = &clusterClassVariables[i]
+	}
```

Comment:
nit: this code seems repeated, let's create two small util func

Reply:
> this code seems repeated
it is, although I personally prefer having it explicit as it's easier to read vs when it's hidden behind a util func.

But I"m fine with moving it to a shared util.